File: examples/ruby/src/renderer.rb
# frozen_string_literal: true

require_relative "shape"

# Collects shapes and renders a summary — showcases Enumerable, blocks, and Struct.
class Renderer
  include Enumerable

  Stats = Struct.new(:count, :total_area, :largest, keyword_init: true)

  def initialize
    @shapes = []
  end

  def <<(shape)
    @shapes << shape
    self
  end

  def each(&block)
    @shapes.each(&block)
  end

  def stats
    Stats.new(
      count: @shapes.size,
      total_area: @shapes.sum(&:area),
      largest: @shapes.max_by(&:area)
    )
  end

  def render
    puts "┌#{'─' * 40}┐"
    puts "│ Scene: #{@shapes.size} shapes#{' ' * (31 - @shapes.size.to_s.length)}│"
    puts "├#{'─' * 40}┤"

    @shapes.each_with_index do |shape, i|
      line = "  #{i + 1}. #{shape} — area #{format('%.2f', shape.area)}"
      puts "│ #{line.ljust(38)} │"
    end

    puts "├#{'─' * 40}┤"

    s = stats
    summary = "  Total area: #{format('%.2f', s.total_area)}"
    puts "│ #{summary.ljust(38)} │"
    puts "└#{'─' * 40}┘"
  end
end

File: examples/rust/src/renderer.rs
use crate::shape::Shape;
use std::io::Write;

pub struct Renderer {
    shapes: Vec<Box<dyn Shape>>,
}

impl Renderer {
    pub fn new() -> Self {
        Self { shapes: Vec::new() }
    }

    pub fn add(&mut self, shape: Box<dyn Shape>) {
        self.shapes.push(shape);
    }

    pub fn render(&self, out: &mut dyn Write) {
        writeln!(out, "=== Rendering {} shape(s) ===", self.shapes.len()).unwrap();
        for s in &self.shapes {
            writeln!(out, "{}\n", s.describe()).unwrap();
        }
        writeln!(out, "Total area: {:.2}", self.total_area()).unwrap();
    }

    pub fn total_area(&self) -> f64 {
        self.shapes.iter().map(|s| s.area()).sum()
    }

    pub fn count(&self) -> usize {
        self.shapes.len()
    }

    pub fn filter<F: Fn(&dyn Shape) -> bool>(&self, pred: F) -> Vec<&dyn Shape> {
        self.shapes.iter()
            .map(|s| s.as_ref())
            .filter(|s| pred(*s))
            .collect()
    }
}

impl Default for Renderer {
    fn default() -> Self {
        Self::new()
    }
}
